libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -1497,11 +1497,21 @@ void AP_GPS::handle_gps_rtcm_fragment(uint8_t flags, const uint8_t *data, uint8_
 
     const uint8_t fragment = (flags >> 1U) & 0x03;
     const uint8_t sequence = (flags >> 3U) & 0x1F;
+    uint8_t* start_of_fragment_in_buffer =
+        &rtcm_buffer->buffer[MAVLINK_MSG_GPS_RTCM_DATA_FIELD_DATA_LEN * (uint16_t)fragment];
+    bool should_clear_previous_fragments = false;
 
-    // see if this fragment is consistent with existing fragments
-    if (rtcm_buffer->fragments_received &&
-        (rtcm_buffer->sequence != sequence ||
-         (rtcm_buffer->fragments_received & (1U<<fragment)))) {
+    if (rtcm_buffer->fragments_received) {
+        const bool sequence_nr_changed = rtcm_buffer->sequence != sequence;
+        const bool seen_this_fragment_index = rtcm_buffer->fragments_received & (1U << fragment);
+        const bool duplicate_fragment = seen_this_fragment_index && (
```

Comment:
This logic is much more convoluted then it needs to be. The factoring to the `const bool`s leads to using `seen_this_fragment_index` twice which is unneeded. Why not just add the `memcmp` to the original if block?

Alternatively if we just `memcmp`d to declare it to be a duplicate buffer entry we shouldn't need to `memcpy` it or do any of the buffer checking below, we should be able to just immediately return from the function, which also ameliorates the cost of doing the `memcmp`.

Reply:
Added early exit when a duplicate was detected; this also alleviates the need for a separate `duplicate_fragment` variable.